File: 101/124. Binary Tree Maximum Path Sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxPathSum = function (root) {

    let maxV = Number.MIN_SAFE_INTEGER

    const dfs = (root) => {
        if (!root) {
            return 0
        }
        const left = Math.max(0, dfs(root.left))
        const right = Math.max(0, dfs(root.right))
        maxV = Math.max(maxV, left + right + root.val)
        return Math.max(left, right) + root.val
    }
    dfs(root)
    return maxV
};

File: 101/124. Binary Tree Maximum Path Sum.go
package main

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxPathSum(root *TreeNode) int {

	var dfs func(root *TreeNode) int
	max := math.MinInt
	dfs = func(root *TreeNode) int {
		if root == nil {
			return 0
		}

		left := Max(0, dfs(root.Left))
		right := Max(0, dfs(root.Right))
		max = Max(max, left+right+root.Val)
		return Max(left, right) + root.Val
	}
	dfs(root)
	return max
}

func Max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
